updated for GNU Midnight Commander 4.8.25-68-g356672787

This patch is to fix an issue in mc (midnight commander)
whereby sometimes the subshell output was truncated and if small enough
completely discarded.
The fix handles the case when the order of two different inputs
(shell output and the pwd from PROMPT_COMMAND) don't arrive in that order.
This happens sometimes and this is when the truncation was in effect.
Also note that shell being STOP-ped can also happen in any order with the above two.

reported issue: https://github.com/MidnightCommander/mc/issues/44#issuecomment-60467020

UPdate: subshell.c location changed with commit: 974ab368ec30fcd68e595252668d1542a5352619

Index: /var/tmp/portage/app-misc/mc-9999/work/mc-9999/src/subshell/common.c
===================================================================
--- orig/src/subshell/common.c
+++ mod/src/subshell/common.c
@@ -739,20 +739,19 @@ feed_subshell (int how, gboolean fail_on
     int i;                      /* Loop counter */
 
     struct timeval wtime;       /* Maximum time we wait for the subshell */
-    struct timeval *wptr;
+    int gotpwdoutput=FALSE;
 
     should_read_new_subshell_prompt = FALSE;
     subshell_should_clear_command_line = FALSE;
 
-    /* we wait up to 10 seconds if fail_on_error, forever otherwise */
-    wtime.tv_sec = 10;
-    wtime.tv_usec = 0;
-    wptr = fail_on_error ? &wtime : NULL;
 
     while (TRUE)
     {
         int maxfdp;
 
+        wtime.tv_sec = (fail_on_error && !gotpwdoutput) ? 10:0;
+        wtime.tv_usec = 10000;
+
         if (!subshell_alive)
             return FALSE;
 
@@ -768,7 +767,7 @@ feed_subshell (int how, gboolean fail_on
             maxfdp = MAX (maxfdp, STDIN_FILENO);
         }
 
-        if (select (maxfdp + 1, &read_set, NULL, NULL, wptr) == -1)
+        if (select (maxfdp + 1, &read_set, NULL, NULL, ( (ACTIVE == subshell_state && subshell_ready) ? NULL : &wtime) ) == -1)
         {
             /* Despite using SA_RESTART, we still have to check for this */
             if (errno == EINTR)
@@ -786,11 +785,6 @@ feed_subshell (int how, gboolean fail_on
 
         if (FD_ISSET (mc_global.tty.subshell_pty, &read_set))
             /* Read from the subshell, write to stdout */
-
-            /* This loop improves performance by reducing context switches
-               by a factor of 20 or so... unfortunately, it also hangs MC
-               randomly, because of an apparent Linux bug.  Investigate. */
-            /* for (i=0; i<5; ++i)  * FIXME -- experimental */
         {
             bytes = read (mc_global.tty.subshell_pty, pty_buffer, sizeof (pty_buffer));
 
@@ -831,16 +825,7 @@ feed_subshell (int how, gboolean fail_on
 
             subshell_cwd[bytes - 1] = '\0';     /* Squash the final '\n' */
 
-            synchronize ();
-
-            clear_subshell_prompt_string ();
-            should_read_new_subshell_prompt = TRUE;
-            subshell_ready = TRUE;
-            if (subshell_state == RUNNING_COMMAND)
-            {
-                subshell_state = INACTIVE;
-                return TRUE;
-            }
+            gotpwdoutput=TRUE;
         }
 
         else if (FD_ISSET (STDIN_FILENO, &read_set))
@@ -888,9 +873,35 @@ feed_subshell (int how, gboolean fail_on
                 subshell_ready = FALSE;
             }
         }
-        else
-            return FALSE;
+        else {
+          if (fail_on_error) {
+            if (subshell_stopped && gotpwdoutput && subshell_state==RUNNING_COMMAND) {
+              break;
+            }else{
+              return FALSE;
+            }
+          }else {
+            if (subshell_stopped) {
+              if (subshell_state == ACTIVE) {
+                synchronize ();
+                subshell_ready = TRUE;
+              } else {
+                break;
+              }
+            }
+          }
+        }
+    }
+
+    synchronize ();
+    clear_subshell_prompt_string ();
+    should_read_new_subshell_prompt = TRUE;
+    subshell_ready = TRUE;
+    if (subshell_state == RUNNING_COMMAND)
+    {
+      subshell_state = INACTIVE;
     }
+    return TRUE;
 }
 
 /* --------------------------------------------------------------------------------------------- */
